File: reference_engine/economics.py
"""Simple and discounted economic calculations with explicit assumptions."""

from __future__ import annotations

import math
from collections.abc import Sequence


class EconomicsError(ValueError):
    """Raised for invalid economic assumptions."""


def annual_energy_cost(
    annual_electricity_kwh: float, electricity_price_per_kwh: float | None
) -> float | None:
    if electricity_price_per_kwh is None:
        return None
    energy = float(annual_electricity_kwh)
    price = float(electricity_price_per_kwh)
    if not math.isfinite(energy) or energy < 0.0:
        raise EconomicsError("Annual electricity must be finite and non-negative")
    if not math.isfinite(price) or price < 0.0:
        raise EconomicsError("Electricity price must be finite and non-negative")
    return energy * price


def simple_payback_years(
    incremental_installed_cost: float,
    annual_operating_cost_saving: float,
    absolute_tolerance: float,
) -> float | None:
    incremental = float(incremental_installed_cost)
    saving = float(annual_operating_cost_saving)
    if incremental <= absolute_tolerance:
        return 0.0
    if saving <= absolute_tolerance:
        return None
    return incremental / saving


def lifecycle_cost(
    installed_cost: float,
    annual_energy_cost_year_one: float,
    annual_maintenance_cost: float,
    analysis_period_years: int,
    discount_rate_fraction: float,
    energy_price_escalation_fraction: float,
    replacements: Sequence[dict[str, float]],
    residual_value: float,
) -> float:
    years = int(analysis_period_years)
    rate = float(discount_rate_fraction)
    escalation = float(energy_price_escalation_fraction)
    if years < 0:
        raise EconomicsError("analysis_period_years must be non-negative")
    if rate <= -1.0 or escalation <= -1.0:
        raise EconomicsError("Discount and escalation fractions must exceed -1")
    replacement_by_year: dict[int, float] = {}
    for replacement in replacements:
        year = int(replacement["year"])
        cost = float(replacement["cost"])
        if year < 1 or year > years:
            raise EconomicsError("Replacement year must be inside the analysis period")
        replacement_by_year[year] = replacement_by_year.get(year, 0.0) + cost
    total = float(installed_cost)
    for year in range(1, years + 1):
        energy = float(annual_energy_cost_year_one) * (1.0 + escalation) ** (year - 1)
        annual = energy + float(annual_maintenance_cost) + replacement_by_year.get(year, 0.0)
        total += annual / (1.0 + rate) ** year
    total -= float(residual_value) / (1.0 + rate) ** years if years else float(residual_value)
    return total


def compare_lifecycle_costs(
    gshp_installed_cost: float | None,
    ashp_installed_cost: float | None,
    gshp_annual_energy_cost: float | None,
    ashp_annual_energy_cost: float | None,
    gshp_annual_maintenance_cost: float,
    ashp_annual_maintenance_cost: float,
    analysis_period_years: int,
    discount_rate_fraction: float,
    energy_price_escalation_fraction: float,
    gshp_replacements: Sequence[dict[str, float]],
    ashp_replacements: Sequence[dict[str, float]],
    gshp_residual_value: float,
    ashp_residual_value: float,
    absolute_tolerance: float,
) -> dict[str, float | None]:
    if None in (
        gshp_installed_cost,
        ashp_installed_cost,
        gshp_annual_energy_cost,
        ashp_annual_energy_cost,
    ):
        return {
            "incremental_installed_cost": None,
            "annual_operating_cost_saving": None,
            "simple_payback_years": None,
            "gshp_lifecycle_cost": None,
            "ashp_lifecycle_cost": None,
            "npv_of_gshp_choice": None,
        }
    incremental = float(gshp_installed_cost) - float(ashp_installed_cost)
    annual_saving = (
        float(ashp_annual_energy_cost)
        + float(ashp_annual_maintenance_cost)
        - float(gshp_annual_energy_cost)
        - float(gshp_annual_maintenance_cost)
    )
    gshp_lcc = lifecycle_cost(
        float(gshp_installed_cost),
        float(gshp_annual_energy_cost),
        gshp_annual_maintenance_cost,
        analysis_period_years,
        discount_rate_fraction,
        energy_price_escalation_fraction,
        gshp_replacements,
        gshp_residual_value,
    )
    ashp_lcc = lifecycle_cost(
        float(ashp_installed_cost),
        float(ashp_annual_energy_cost),
        ashp_annual_maintenance_cost,
        analysis_period_years,
        discount_rate_fraction,
        energy_price_escalation_fraction,
        ashp_replacements,
        ashp_residual_value,
    )
    return {
        "incremental_installed_cost": incremental,
        "annual_operating_cost_saving": annual_saving,
        "simple_payback_years": simple_payback_years(
            incremental, annual_saving, absolute_tolerance
        ),
        "gshp_lifecycle_cost": gshp_lcc,
        "ashp_lifecycle_cost": ashp_lcc,
        "npv_of_gshp_choice": ashp_lcc - gshp_lcc,
    }

File: src/engine/economics.ts
import type { EconomicsParameters, Replacement } from "../parameters/types";
import { EconomicsError } from "./errors";
import type { EconomicsResult } from "./types";

export function annualEnergyCost(
  annualElectricityKwh: number,
  electricityPricePerKwh: number | null,
): number | null {
  if (electricityPricePerKwh === null) return null;
  if (!Number.isFinite(annualElectricityKwh) || annualElectricityKwh < 0) {
    throw new EconomicsError("Annual electricity must be finite and non-negative");
  }
  if (!Number.isFinite(electricityPricePerKwh) || electricityPricePerKwh < 0) {
    throw new EconomicsError("Electricity price must be finite and non-negative");
  }
  return annualElectricityKwh * electricityPricePerKwh;
}

export function simplePaybackYears(
  incrementalInstalledCost: number,
  annualOperatingCostSaving: number,
  absoluteTolerance: number,
): number | null {
  if (incrementalInstalledCost <= absoluteTolerance) return 0;
  if (annualOperatingCostSaving <= absoluteTolerance) return null;
  return incrementalInstalledCost / annualOperatingCostSaving;
}

export function lifecycleCost(
  installedCost: number,
  annualEnergyCostYearOne: number,
  annualMaintenanceCost: number,
  analysisPeriodYears: number,
  discountRateFraction: number,
  energyPriceEscalationFraction: number,
  replacements: readonly Replacement[],
  residualValue: number,
): number {
  const years = Math.trunc(analysisPeriodYears);
  if (years < 0) throw new EconomicsError("analysisPeriodYears must be non-negative");
  if (discountRateFraction <= -1 || energyPriceEscalationFraction <= -1) {
    throw new EconomicsError("Discount and escalation fractions must exceed -1");
  }
  const replacementByYear = new Map<number, number>();
  for (const replacement of replacements) {
    const year = Math.trunc(replacement.year);
    if (year < 1 || year > years) {
      throw new EconomicsError("Replacement year must be inside the analysis period");
    }
    replacementByYear.set(year, (replacementByYear.get(year) ?? 0) + replacement.cost);
  }
  let total = installedCost;
  for (let year = 1; year <= years; year += 1) {
    const energy = annualEnergyCostYearOne * (1 + energyPriceEscalationFraction) ** (year - 1);
    const annual = energy + annualMaintenanceCost + (replacementByYear.get(year) ?? 0);
    total += annual / (1 + discountRateFraction) ** year;
  }
  total -= years === 0
    ? residualValue
    : residualValue / (1 + discountRateFraction) ** years;
  return total;
}

export function compareLifecycleCosts(
  parameters: EconomicsParameters,
  gshpAnnualEnergyCost: number | null,
  ashpAnnualEnergyCost: number | null,
  absoluteTolerance: number,
): EconomicsResult {
  if (
    parameters.gshp_installed_cost === null ||
    parameters.ashp_installed_cost === null ||
    gshpAnnualEnergyCost === null ||
    ashpAnnualEnergyCost === null
  ) {
    return {
      incrementalInstalledCost: null,
      annualOperatingCostSaving: null,
      simplePaybackYears: null,
      gshpLifecycleCost: null,
      ashpLifecycleCost: null,
      npvOfGshpChoice: null,
    };
  }
  const incrementalInstalledCost =
    parameters.gshp_installed_cost - parameters.ashp_installed_cost;
  const annualOperatingCostSaving =
    ashpAnnualEnergyCost + parameters.ashp_annual_maintenance_cost -
    gshpAnnualEnergyCost - parameters.gshp_annual_maintenance_cost;
  const gshpLifecycleCost = lifecycleCost(
    parameters.gshp_installed_cost,
    gshpAnnualEnergyCost,
    parameters.gshp_annual_maintenance_cost,
    parameters.analysis_period_years,
    parameters.discount_rate_fraction,
    parameters.electricity_price_escalation_fraction,
    parameters.gshp_replacements,
    parameters.gshp_residual_value,
  );
  const ashpLifecycleCost = lifecycleCost(
    parameters.ashp_installed_cost,
    ashpAnnualEnergyCost,
    parameters.ashp_annual_maintenance_cost,
    parameters.analysis_period_years,
    parameters.discount_rate_fraction,
    parameters.electricity_price_escalation_fraction,
    parameters.ashp_replacements,
    parameters.ashp_residual_value,
  );
  return {
    incrementalInstalledCost,
    annualOperatingCostSaving,
    simplePaybackYears: simplePaybackYears(
      incrementalInstalledCost,
      annualOperatingCostSaving,
      absoluteTolerance,
    ),
    gshpLifecycleCost,
    ashpLifecycleCost,
    npvOfGshpChoice: ashpLifecycleCost - gshpLifecycleCost,
  };
}
